python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -411,26 +419,43 @@ def write(self, value: Any):
         else:
             serialized_value = value
 
-        self._resize_channel_if_needed(serialized_value)
+        try:
+            start_time = time.monotonic()
+            self._resize_channel_if_needed(serialized_value, timeout_ms)
+            if timeout is not None:
+                timeout_ms -= int((time.monotonic() - start_time) * 1000)
+                if timeout_ms <= 0:
+                    raise TimeoutError(f"Timed out when writing to {self}.")
```

Comment:
Let's define our own timeout error that inherits from TimeoutError and RayChannelError.